### PR TITLE
feat: add call bundle

### PR DESCRIPTION
**Pull request by mattsse**
adds call bundle implementation ported from 

https://github.com/flashbots/builder/blob/03ee71cf0a344397204f65ff6d3a917ee8e06724/internal/ethapi/api.go#L2212-L2212

this ships it as standalone function first.
need to decide if and how we want to enable this via cli

### CODE REVIEW

File: crates/rpc/rpc-types/src/mev.rs [hunk 1]
```diff
@@ -641,9 +641,16 @@ pub struct EthCallBundleTransactionResult {
...
     pub tx_hash: B256,
-    pub value: Bytes,
+    /// Contains the return data if the transaction succeeded
+    ///
+    /// Note: this is mutually exclusive with `revert`
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub value: Option<Bytes>,
+    /// Contains the return data if the transaction reverted
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub revert: Option<Bytes>,
```
Comment (gakonst) [inline crates/rpc/rpc-types/src/mev.rs:653]: should this be a tagged union enum ?

File: crates/rpc/rpc/src/eth/bundle.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                let mut hash_bytes = Vec::with_capacity(32 * transactions.len());
+
+                let mut evm = revm::EVM::with_env(env);
+                evm.database(db);
+
+                let mut results = Vec::with_capacity(transactions.len());
+                let mut transactions = transactions.into_iter().peekable();
+
+                while let Some(tx) = transactions.next() {
+                    let tx = tx.into_ecrecovered_transaction();
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/bundle.rs:95]: perf: can do the recovery outside of the loop in parallel

Comment (mattsse) [reply crates/rpc/rpc/src/eth/bundle.rs:95]: I don't think it is worth it considering that we expect a lot of concurrent requests.

File: crates/rpc/rpc/src/eth/bundle.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            .spawn_with_state_at_block(at, move |state| {
+                let coinbase = block_env.coinbase;
+                let basefee = Some(block_env.basefee.to::<u64>());
+                let env = Env { cfg, block: block_env, tx: TxEnv::default() };
+                let db = CacheDB::new(StateProviderDatabase::new(state));
+
+                let initial_coinbase =
+                    DatabaseRef::basic(&db, coinbase)?.map(|acc| acc.balance).unwrap_or_default();
+                let mut coinbasebalance_before_tx = initial_coinbase;
+                let mut coinbasebalance_after_tx = initial_coinbase;
```
Comment (Rjected) [inline crates/rpc/rpc/src/eth/bundle.rs:83]: nit: rename to `coinbase_balance_before_tx` and `coinbase_balance_after_tx`

File: crates/rpc/rpc-types/src/mev.rs [hunk 4]
```diff
@@ -613,9 +613,9 @@ pub struct EthCallBundle {
 #[derive(Default, Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
 #[serde(rename_all = "camelCase")]
 pub struct EthCallBundleResponse {
+    pub bundle_hash: String,
```
Comment (Rjected) [inline crates/rpc/rpc-types/src/mev.rs:616]: any reason why this shouldn't be a H256?

**Review by Rjected — APPROVED** [review]
lgtm so far after reading the builder code, aside from nits

File: crates/rpc/rpc-types/src/mev.rs (hunk 4 above)
Comment (mattsse) [reply crates/rpc/rpc-types/src/mev.rs:616]: I believe there's no spec on how this computed, but I believe it's always B256

File: crates/rpc/rpc/src/eth/bundle.rs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                    if transactions.peek().is_some() {
+                        // need to apply the state changes of this call before executing
+                        // the next call
+                        evm.db.as_mut().expect("is set").commit(state)
+                    }
+                }
+
+                // populate the response
+
+                let coinbase_diff = initial_coinbase.saturating_sub(coinbase_balance_after_tx);
```
Comment (rkrasiuk) [inline crates/rpc/rpc/src/eth/bundle.rs:153]: i've made this mistake so many times as well :) should be inversed

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc/src/eth/bundle.rs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+        // need to adjust the timestamp for the next block
+        if let Some(timestamp) = timestamp {
+            block_env.timestamp = U256::from(timestamp);
+        } else {
+            block_env.timestamp += U256::from(12);
+        }
+
+        // use the block number of the request
+        block_env.number = U256::from(block_number);
```
Comment (rkrasiuk) [inline crates/rpc/rpc/src/eth/bundle.rs:70]: should we validate that `block_number == state_block_number + 1` or is it not required?

Comment (mattsse) [reply crates/rpc/rpc/src/eth/bundle.rs:70]: there's no such check in the ref impl

**Review by rkrasiuk — APPROVED** [review]
lgtm

**Pull request merged**